sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java

```diff
@@ -664,30 +667,33 @@ Timestamps snapshotTimestamps() {
 
     private RntbdRequestRecord addPendingRequestRecord(final ChannelHandlerContext context, final RntbdRequestRecord record) {
 
-        return this.pendingRequests.compute(record.transportRequestId(), (id, current) -> {
+        AtomicReference<Timeout> pendingRequestTimeout = new AtomicReference<>();
+        this.pendingRequests.compute(record.transportRequestId(), (id, current) -> {
 
             reportIssueUnless(current == null, context, "id: {}, current: {}, request: {}", record);
             record.pendingRequestQueueSize(pendingRequests.size());
 
-            final Timeout pendingRequestTimeout = record.newTimeout(timeout -> {
+            pendingRequestTimeout.set(record.newTimeout(timeout -> {
 
                 // We don't wish to complete on the timeout thread, but rather on a thread doled out by our executor
                 requestExpirationExecutor.execute(record::expire);
-            });
-
-            record.whenComplete((response, error) -> {
-                this.pendingRequests.remove(id);
-                pendingRequestTimeout.cancel();
-            });
+            }));
 
             return record;
+        });
 
+        record.whenComplete((response, error) -> {
+            this.pendingRequests.remove(record.transportRequestId());
```

Comment:
This is great finding @xinlian12 
May be some comments on this will be good to add.

Reply:
Added comments